### DISCUSSION TITLE
Messaging: working on Android but not on iOS

### DISCUSSION THREAD

**Discussion by eliorivero** [invertase/react-native-firebase#4151] [Q&A] [answered]
Hi! I'm working on receiving data only FCM messages. I was able to work with `onMessage()` and `setBackgroundMessageHandler()` and receive messages on Android.

However, I'm not getting anything on iOS.

I went through all the steps to grant permissions and integrate it as described in the documentation. I'm using a React Native Firebase [PHP library to handle messaging](https://firebase-php.readthedocs.io/en/stable/cloud-messaging.html#sending-a-fully-configured-raw-message) and passing `apns-priority` 10 as header and `content-available` 1 as payload as pointed by FCM docs. The request is working fine and the token is correct (I tested with a fake token and got an error saying that it was an unknown entity).

I've set `FirebaseAppDelegateProxyEnabled` to 0 (boolean) in `Info.plist`, I also tried NO (string), same thing. When I don't set this or when I explicitly map the APN token to the FCM registration token, I get an error saying _Invalid registration token_.

Will deeply appreciate any help. Thanks!

**Comment by mikehardy** [comment, marked as answer]
Hi there! There are (were?) lots of open issues about this. I don't think you should do anything in the plist. I would revert all of that.

You may need to send content-available: 1 in your JSON.

Best practice for background messaging on iOS:

- use a real device at all times, it does not work on the simulator
- test the release version of your app if you can (there are timing issues involved, performance delay of debug bundle download/parse/execute may be important and be too slow or cause iOS to throttle the app and not deliver)
- launch it from Xcode (you can watch the logs this way to see if your device is refusing to wake the app because of throttling or timeouts etc)
- send the JSON from the command line so you can control it carefully
- make sure content-available is in there and set to 1
- make sure your background message handler is an async function and set in before app is registered

That should be it? Although a triple-check of entitlements and configs and everything is perhaps helpful

**Reply by andersonaddo** [reply]
More info on the headers you'll need:
https://github.com/invertase/react-native-firebase/issues/3999#issuecomment-664031711

**Reply by eliorivero** [reply]
Thank you!

**Comment by eliorivero** [comment]
Removing the `FirebaseAppDelegateProxyEnabled` from `Info.plist` did the trick!
I'm still not getting them when the app is in the background but I'll consider this answered because my issue was that I wasn't getting anything at all on iOS.
Thanks!

**Reply by Salakar** [reply]
Was there some docs somewhere that suggested using `FirebaseAppDelegateProxyEnabled`? Changing this would break messaging, some auth methods & dynamic links - so just want to make sure it's not something we've said in our docs to do

**Reply by eliorivero** [reply]
Not in the docs. I got it from somewhere, while trying a bunch of things to make it work.

**Reply by Stas-Buzunko** [reply]
it's from xcode logs
```
[Firebase/Messaging][I-FCM001000] FIRMessaging Remote Notifications proxy enabled, will swizzle remote notification receiver handlers. If you'd prefer to manually integrate Firebase Messaging, add "FirebaseAppDelegateProxyEnabled" to your Info.plist, and set it to NO. Follow the instructions at:
https://firebase.google.com/docs/cloud-messaging/ios/client#method_swizzling_in_firebase_messaging
to ensure proper integration.
```

**Reply by mikehardy** [reply]
Ah, that makes sense. This is not something we recommend doing at all though in practice. It's open source, and this is a complicated topic both in the area of business requirements, existing apps, library ecosystems etc etc but for react-native-firebase to work correctly out of the box we do need AppDelegate swizzling or all of the iOS "app responds to external events" things won't work without manual configurations as Salakar mentioned. And with current react-native-firebase versions at least I receive background data-only messages on iOS as reliably *as I expect* (that is, when background data is allowed, when power save is not on, when I haven't sent too many, when the app has been used semi-frequently so it gets power budget, when it hasn't been force-closed). My project sends 3 messages a day at specific times and they arrive and open the app like a clock in the right conditions.
